### PR TITLE
Add ca-ensure-requirements command

### PR DESCRIPTION
**Pull request by elliotthilaire-ca**
## Purpose 

This introduces a ca-ensure-requirements command to the ca-common devbox plugin.

ca-ensure-requirements will sit at the front of `verify` and `check` commands step and guarantees its prerequisite processes are actually up and healthy.

Those prerequisites can be shaped in different ways, fail in different ways, and be encountered in different starting states. Without a shared plugin we'll end up with adhoc implementations that diverge in assumptions and behaviour. Independant implementations code to handle requirements for verify will also most likely not be tested and contain unpredictable edge case bugs.

| Dimension | What it varies | Values |
  |---|---|---|
  | **Requirement shape** | How `verify`'s prerequisites are expressed | • depends on a long-running daemon process<br>• depends on one-shot commands only |
  | **Health outcome** | What happens when the requirements run | • everything is green<br>• a process command exits with a failure<br>• a process fails to become ready |
  | **Starting state** | State of `devbox services up` when we arrive | • already running<br>• not running<br>• running but missing some of verify's requirements |

## Notes for reviewers

If you're not in the mood to review hundreds of lines of bash I'd really appreciate checking out the example branches mentioned below trying the using devbox run verify in a few different scenarios. See if you can find edge cases and break the command. Or try this out in a project you know well.

CI timeout/failure is fixed in this PR: https://github.com/cultureamp/devbox-extras/pull/91

## Verification

Example of this being used in a project with services:  https://github.com/cultureamp/bef-rails-pg-demo/pull/1061

Example of this being used in a project without services: https://github.com/cultureamp/cdk-constructs/pull/2437

### python server for testing

Seems like overkill tried to find a better option but did not. Open to suggestions.

### Naming the command ca-ensure-processes

This plugin is specific to using process compose so maybe that name will be better. However people treat services and processes and synonymous where this can also be used for one-shot processes. 

### Future of run-with-services-up.

run-with-services-up now duplicates a handful of shared functionality. A follow up to this should be to extract common logic into the _lib.bash file. Alternatively we may be able to deprecate it and use this instead.

## Additional notes

Claude thinks it's picked up a bug in devbox causing file descriptor leaks. This was found when it was debugging tests hanging. The code contains a work around for this. As a followup we can see if this is a real issue and can be solved in devbox.

### CODE REVIEW

File: plugins/ca-common/bin/_lib.bash [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    fi
+    echo "::endgroup::"
+    echo ""
+  done
+}
+
+# Echo "yes" if any closure member in $yaml_file looks daemon-shaped
+# (has readiness_probe or is_daemon: true), otherwise "no". Reads YAML
+# directly via ruby because the API may not be up yet. Ruby's yaml stdlib
+# ships with the system ruby on macOS; for project devboxes ruby is in PATH.
```
Comment (elliotthilaire-ca) [inline plugins/ca-common/bin/_lib.bash]: We can't depend on Ruby being installed. Ruby is available on MacOS but not everywhere this will be used.

File: plugins/ca-common/bin/ca-ensure-requirements [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+  fi
+
+  if [ -z "$pcfile" ]; then
+    pcfile="$DEVBOX_PROJECT_ROOT/process-compose.yaml"
+    if [ ! -f "$pcfile" ] && [ -f "$DEVBOX_PROJECT_ROOT/process-compose.yml" ]; then
+      pcfile="$DEVBOX_PROJECT_ROOT/process-compose.yml"
+    fi
+  fi
+
+  if [ "$#" -ne 1 ]; then
```
[the remark this answers is not in the document]
Comment (elliotthilaire-ca) [reply plugins/ca-common/bin/ca-ensure-requirements:106]: Implemented.

**Review by jay-aye-see-kay — APPROVED** [review]
